src/dune_rules/artifact_substitution.ml

```diff
@@ -48,8 +48,22 @@ type conf =
   ; get_location : Section.t -> Package.Name.t -> Path.t
   ; get_config_path : configpath -> Path.t option
   ; hardcoded_ocaml_path : hardcoded_ocaml_path
+  ; sign_hook : (Path.t -> unit Fiber.t) option
   }
 
+let mac_codesign_hook ~codesign path =
+  Process.run Strict codesign [ "--sign"; "-"; Path.to_string path ]
+
+let sign_hook_of_context (context : Context.t) =
+  let config = context.ocaml_config in
+  match (Ocaml_config.system config, Ocaml_config.architecture config) with
+  | "macosx", "arm64" -> (
+    let codesign_name = "codesign" in
+    match Bin.which ~path:context.path codesign_name with
+    | None -> Utils.program_not_found ~loc:None codesign_name
```

Comment:
Could you add an appropriate hint to the error? To tell the user why we need this binary and where it comes from (xcode?).

Reply:
as I understand it, it should be part of the base system. I'll add a short hint.